File: solutions/code-carl/ch06_tree/T11_LC0110_IsBalanced.java
/**
 * Practice template for:
 * - LeetCode 110. Balanced Binary Tree
 *   https://leetcode.cn/problems/balanced-binary-tree/
 */
public class T11_LC0110_IsBalanced {
    public boolean isBalanced(TreeNode root) {
        return getHeight(root) != -1;
    }

    public int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftH = getHeight(root.left);
        if (leftH == -1) {
            return -1;
        }
        int rightH = getHeight(root.right);
        if (rightH == -1) {
            return -1;
        }
        if (Math.abs(leftH - rightH) > 1) {
            return -1;
        }

        return Math.max(leftH, rightH) + 1;
    }

}

File: solutions/code-carl/ch06_tree/t11_lc0110_is_balanced.py
"""LeetCode 110. Balanced Binary Tree
https://leetcode.cn/problems/balanced-binary-tree/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        left =  self.getHeight(root.left)
        right = self.getHeight(root.right)
        return left != -1 and right != -1 and abs(left - right) <= 1

    def getHeight(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        left = self.getHeight(root.left)
        if left == -1:
            return -1

        right = self.getHeight(root.right)
        if right == -1:
            return -1

        if abs(left - right) > 1:
            return -1

        return max(left, right) + 1


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert solution.isBalanced(root1) is True

    root2 = TreeNode(
        1, TreeNode(2, TreeNode(3, TreeNode(4), TreeNode(4)), TreeNode(3)), TreeNode(2)
    )
    assert solution.isBalanced(root2) is False

    print("All tests passed.")
